### ISSUE TITLE
Wrong variable binding and missed matching in OR `match` expressions 

### ISSUE THREAD

**Issue by ironcev** [FuelLabs/sway#5122]
Consider the following code and the asserts that _all assert to true_:

```Sway
script;

struct Struct {  x: u64, y: u64, z: u64, }

fn main() {
    let s = Struct { x: 1, y: 2, z: 3, };

    let a = match s {
        Struct { x, y: 2, z: 3 } | Struct { x: 0, y: x, z: 0 } | Struct { x: 0, y: 0, z: x } => x,
        _ => 1111,
    };
    
    assert(a == 3); // BUG: Should be 1.

    let a = match s {
        Struct { x, y: 0, z: 0 } | Struct { x: 1, y: x, z: 3 } | Struct { x: 0, y: 0, z: x } => x,
        _ => 1111,
    };
    
    assert(a == 1111); // BUG: Should be 2.

    let a = match s {
        Struct { x, y: 0, z: 0 } | Struct { x: 0, y: x, z: 0 } | Struct { x: 1, y: 2, z: x } => x,
        _ => 1111,
    };

    assert(a == 1111); // BUG: Should be 3.

    let a = match s {
        Struct { x, y: 0, z: 0 } | Struct { x: 0, y: x, z: 0 } | Struct { x: 0, y: 0, z: x } => x,
        _ => 1111,
    };
    
    assert(a == 1111); // OK.
}
```
The reason for the first bug (3 instead of 1) is in the design of the [matcher](https://github.com/FuelLabs/sway/blob/44f46a0a905e3f99b40c2f0019f31ba487e20cb0/sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs#L173) that returns only the [MatchDeclMap](https://github.com/FuelLabs/sway/blob/44f46a0a905e3f99b40c2f0019f31ba487e20cb0/sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs#L24) of the last variant.

In other words, no matter which of the variants gets matched the value for the variable will always be extracted from the last variant.

Why the match-arm does not match at all in the other two examples is not immediately clear.

The consequence of this issue on the current design are:
- we need to change the [MatcherResult](https://github.com/FuelLabs/sway/blob/44f46a0a905e3f99b40c2f0019f31ba487e20cb0/sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs#L26) structure to return all `MatchDeclMap`s for all the alternatives.
- we need to change the desugared code to track which of the variants gets matched and to initialize the match-arm variable accordingly.

**Comment by ironcev** [comment]
> Therefore, the only proper thing to do at runtime is to revert.

Beware that we desugar `if-let`s to match expressions. In such cases the final else is not a dummy one and must not revert.
